contracts/auctions/FungibleAuction.sol

```diff
@@ -0,0 +1,165 @@
+    function buy(uint256 amount) external override {
+        require(isActive(), "Auction: not active");
+
+        uint256 availableAmount = sellTokenAddress.balanceOf(address(this));
+
+        require(availableAmount > 0, "Auction: sold out");
+
+        // Allow buyer to buy whatever is left.
+        if (amount > availableAmount) {
+            amount = availableAmount;
+        }
+
+        // The number of tokens required to buy the requested amount (or what is left of it).
+        uint256 buyPrice = getPriceInBuyTokens(amount);
+
+        require(
+            buyTokenAddress.allowance(msg.sender, address(this)) >= buyPrice,
+            "Auction: Auction allowance < tokensToBuy"
+        );
+        require(buyTokenAddress.balanceOf(msg.sender) >= buyPrice, "Auction: Bidder balance < tokensToBuy");
+
+        // Set lowest sell price, used when closing the auction in the auction house.
+        lowestSellPrice = getPrice();
+
+        // Token transfers.
+        buyTokenAddress.safeTransferFrom(msg.sender, address(this), buyPrice);
+        sellTokenAddress.safeTransfer(msg.sender, amount);
+
+        emit Buy(address(sellTokenAddress), amount, lowestSellPrice);
+    }
+
+    /// @dev Withdraw the tokens that were bid.
+    /// TODO: Allow auctioneer to withdraw tokens before auction is finished, in case of long durations?
+    function withdraw() external override {
+        require(msg.sender == auctioneer, "Auction: only auctioneer can withdraw");
+
+        uint256 sellTokenBalance = sellTokenAddress.balanceOf(address(this));
+
+        // Allow withdrawal if all tokens have been sold.
+        if (sellTokenBalance > 0) {
+            require(startTimestamp + duration < block.timestamp, "Auction: auction has not ended yet");
+        }
+
+        buyTokenAddress.safeTransferFrom(address(this), auctioneer, buyTokenAddress.balanceOf(address(this)));
+
+        // Transfer back what has not been sold.
+        if (sellTokenBalance > 0) {
+            sellTokenAddress.safeTransferFrom(address(this), auctioneer, sellTokenBalance);
+        }
+
+        // Callback to auction house for administrative purposes.
+        if (address(auctionHouse) != address(0)) {
+            auctionHouse.closeFungible(
+                address(sellTokenAddress),
+                address(buyTokenAddress),
+                initialAmount - sellTokenBalance,
+                lowestSellPrice
+            );
+        }
+    }
+
+    /// VIEWS
+    /// @dev Get price per token.
+    function getPrice() public view override returns (uint256) {
+        if (block.timestamp <= startTimestamp) {
+            return maxPrice;
+        }
+
+        // Guard against timestamps *very* far into the future and/or long durations.
+        // TODO: Intended? Should an item that can no longer be sold return `minPrice`?
+        if (block.timestamp > startTimestamp.add(duration)) {
+            return minPrice;
```

Comment:
How about returning lowestSellPrice instead? 

Its possible that minPrice is never even hit. In such a case, seems like lowestSellPrice might be a more meaningful number to return.

Reply:
I like the idea!